References
===========


.. [Aibar17]
   Sara Aibar, Carmen Bravo González-Blas, Thomas Moerman, Vân Anh Huynh-Thu, Hana Imrichova, Gert Hulselmans, Florian Rambow, Jean-Christophe Marine, Pierre Geurts, Jan Aerts, Joost van den Oord, Zeynep Kalender Atak, Jasper Wouters & Stein Aerts,
   *SCENIC: single-cell regulatory network inference and clustering*, Nature Methods.

.. [Aran19]
   Dvir Aran, Agnieszka P. Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, Ram P. Naikawadi, Paul J. Wolters, Adam R. Abate, Atul J. Butte & Mallar Bhattacharya,
   *Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage*, Nature Immunology.

.. [Blondel08] 
   Vincent D Blondel, Jean-Loup Guillaume, Renaud Lambiotte, Etienne Lefebvrel, 
   *Fast unfolding of communities in large networks*, J. Stat. Mech..

.. [DeTomaso21]
   David DeTomaso, Nir Yosef,
   *Hotspot identifies informative gene modules across modalities of single-cell genomics*, Cell Systems.

.. [Efremova20]
   Mirjana Efremova, Miquel Vento-Tormo, Sarah A. Teichmann, Roser Vento-Tormo,
   *CellPhoneDB: inferring cell-cell communication from combined expression of multi-subunit ligand-receptor complexes*, Nature Protocols.

.. [Fang23]
   Shuangsang Fang, Mengyang Xu, Lei Cao, Xiaobin Liu, Marija Bezulj, Liwei Tan, Zhiyuan Yuan, Yao Li, Tianyi Xia, Longyu Guo, Vladimir Kovacevic, Junhou Hui, Lidong Guo, Chao Liu, Mengnan Cheng, Li'ang Lin, Zhenbin Wen, Bojana Josic, Nikola Milicevic, Ping Qiu, Qin Lu, Yumei Li, Leying Wang, Luni Hu, Chao Zhang, Qiang Kang, Fengzhen Chen, Ziqing Deng, Junhua Li, Mei Li, Shengkang Li, Yi Zhao, Guangyi Fan, Yong Zhang, Ao Chen, Yuxiang Li, Xun Xu,
   *Stereopy: modeling comparative and spatiotemporal cellular heterogeneity via multi-sample spatial transcriptomics*, bioRxiv.

.. [Hafemeister19] 
   Christoph Hafemeister, Rahul Satija, 
   *Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression*, Genome Biology.

.. [Haghverdi16]
   Laleh Haghverdi, Maren Büttner, F Alexander Wolf, Florian Buettner, Fabian J Theis,
   *Diffusion pseudotime robustly reconstructs lineage branching*, Nature Methods.

.. [Marshall22]
   Jamie L Marshall, Teia Noel, Qingbo S Wang, Haiqi Chen, Evan Murray, Ayshwarya Subramanian, Katherine A Vernon, Silvana Bazua-Valenti, Katie Liguori, Keith Keller, Robert R Stickels, Breanna McBean, Rowan M Heneghan, Astrid Weins, Evan Z Macosko, Fei Chen, Anna Greka,
   *High-resolution Slide-seqV2 spatial transcriptomics enables discovery of disease-specific cell neighborhoods and pathways*, iScience.
   
.. [McInnes18] 
   Leland McInnes, John Healy, James Melville, 
   *UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction*, arXiv.

.. [Satija15] 
   Rahul Satija, Jeffrey A Farrell, David Gennert, Alexander F Schier, Aviv Regev, 
   *Spatial reconstruction of single-cell gene expression data*, Nature Biotechnology.

.. [Shen22] 
   Rongbo Shen, Lin Liu, Zihan Wu, Ying Zhang, Zhiyuan Yuan, Junfu Guo, Fan Yang, Chao Zhang, Bichao Chen, Chao Liu, Jing Guo, Guozhen Fan, Yong Zhang, Yuxiang Li, Xun Xu, Jianhua Yao, 
   *Spatial-ID: a cell typing method for spatially resolved transcriptomics via transfer learning and spatial embedding*, Nature Communications.

.. [Stuart19] 
   Tim Stuart, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck 3rd, Yuhan Hao, Marlon Stoeckius, Peter Smibert, Rahul Satija, 
   *Comprehensive integration of single-cell data*, Cell.

.. [Traag18] 
   Vincent Traag, Ludo Waltman, Nees Jan van Eck, 
   *From Louvain to Leiden: guaranteeing well-connected communities*, Scientific Reports.

.. [VandeSande20]
   Bram Van de Sande, Christopher Flerin, Kristofer Davie, Maxime De Waegeneer, Gert Hulselmans, Sara Aibar, Ruth Seurinck, Wouter Saelens, Robrecht Cannoodt, Quentin Rouchon, Toni Verbeiren, Dries De Maeyer, Joke Reumers, Yvan Saeys & Stein Aerts,
   *A scalable SCENIC workflow for single-cell gene regulatory network analysis*, Nature Protocols.

.. [Wolf19]
   F. Alexander Wolf, Fiona K. Hamey, Mireya Plass, Jordi Solana, Joakim S. Dahlin, Berthold Göttgens, Nikolaus Rajewsky, Lukas Simon, Fabian J. Theis,
   *PAGA: graph abstraction reconciles clustering with trajectory inference through a topology preserving map of single cells*, Genome Biology

.. [Zhang23]
   Chao Zhang, Qiang Kang, Mei Li, Hongqing Xie, Shuangsang Fang, Xun Xu,
   *BatchEval Pipeline: Batch Effect Evaluation Workflow for Multiple Datasets Joint Analysis*, bioRxiv.

.. [Zheng17] 
   Grace X Y Zheng, Jessica M Terry, Phillip Belgrader, Paul Ryvkin, Zachary W Bent, Ryan Wilson, Solongo B Ziraldo, Tobias D Wheeler, Geoff P McDermott, Junjie Zhu, Mark T Gregory, Joe Shuga, Luz Montesclaros, Jason G Underwood, Donald A Masquelier, Stefanie Y Nishimura, Michael Schnall-Levin, Paul W Wyatt, Christopher M Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D Ness, Lan W Beppu, H Joachim Deeg, Christopher McFarland, Keith R Loeb, William J Valente, Nolan G Ericson, Emily A Stevens, Jerald P Radich, Tarjei S Mikkelsen, Benjamin J Hindson, Jason H Bielas, 
   *Massively parallel digital transcriptional profiling of single cells*, Nature Communications.

.. [Xia23]
   Tianyi Xia, Luni Hu, Lulu Zuo, Yunjia Zhang, Mengyang Xu, Qin Lu, Lei Zhang, Lei Cao, Taotao Pan, Bohan Zhang, Bowen Ma, Chuan Chen, Junfu Guo, Chang Shi, Mei Li, Chao Liu, Yuxiang Li, Yong Zhang, Shuangsang Fang,
   *ST-GEARS: Advancing 3D Downstream Research through Accurate Spatial Information Recovery*, bioRxiv.

.. [Bai23]
   Yong Bai, Xiangyu Guo, Keyin Liu, Bingjie Zheng, Yingyue Wang, Qiuhong Luo, Jianhua Yin, Liang Wu, Yuxiang Li, Yong Zhang, Ao Chen, Xun Xu, Xin Jin,
   *Efficient reliability analysis of spatially resolved transcriptomics at varying resolutions using SpaSEG*, bioRxiv.

.. [Shen23]
    Xunan Shen, Ke Huang, Lulu Zuo, Zhongfei Ye, Zeyu Li, Qichao Yu, Xuanxuan Zou, Xiaoyu Wei, Ping Xu, Xin Jin, Xun Xu, Liang Wu, Hongmei Zhu, Pengfei Qin,
    *Inferring cell trajectories of spatial transcriptomics via optimal transport analysis*, bioRxiv.
